client/session.go

```diff
@@ -1548,29 +1590,164 @@ func (s *session) streamBlocksMetadataFromPeer(
 	return nil
 }
 
+// streamBlocksMetadataFromPeerV2 has several heap allocated anonymous
+// function, however, they're only allocated once per peer/shard combination
+// for the entire peer bootstrapping process so performance is acceptable
+func (s *session) streamBlocksMetadataFromPeerV2(
+	namespace ts.ID,
+	shard uint32,
+	peer peer,
+	start, end time.Time,
+	metadataCh chan<- blocksMetadata,
+	progress *streamFromPeersMetrics,
+) error {
+	var (
+		pageToken []byte
+		retrier   = xretry.NewRetrier(xretry.NewOptions().
+				SetBackoffFactor(2).
+				SetMaxRetries(3).
+				SetInitialBackoff(s.streamBlocksMetadataBatchBackoff).
+				SetJitter(true))
+		optionIncludeSizes     = true
+		optionIncludeChecksums = true
+		optionIncludeLastRead  = true
+		moreResults            = true
+	)
+
+	// Declare before loop to avoid redeclaring each iteration
+	attemptFn := func(client rpc.TChanNode) error {
+		tctx, _ := thrift.NewContext(s.streamBlocksMetadataBatchTimeout)
+		req := rpc.NewFetchBlocksMetadataRawV2Request()
+		req.NameSpace = namespace.Data().Get()
+		req.Shard = int32(shard)
+		req.RangeStart = start.UnixNano()
+		req.RangeEnd = end.UnixNano()
+		req.Limit = int64(s.streamBlocksBatchSize)
+		req.PageToken = pageToken
+		req.IncludeSizes = &optionIncludeSizes
+		req.IncludeChecksums = &optionIncludeChecksums
+		req.IncludeLastRead = &optionIncludeLastRead
+
+		progress.metadataFetchBatchCall.Inc(1)
+		result, err := client.FetchBlocksMetadataRawV2(tctx, req)
+		if err != nil {
+			progress.metadataFetchBatchError.Inc(1)
+			return err
+		}
+
+		progress.metadataFetchBatchSuccess.Inc(1)
+		progress.metadataReceived.Inc(int64(len(result.Elements)))
+
+		if result.NextPageToken != nil {
+			// copy will not extend the destination slice automatically
+			nextPageTokenLen := len(result.NextPageToken)
+			pageTokenLen := len(pageToken)
+			delta := nextPageTokenLen - pageTokenLen
+			if delta > 0 {
+				for i := 0; i < delta; i++ {
+					pageToken = append(pageToken, '\x00')
+				}
+			}
+			// Copy the pageToken so we don't have to keep the result around
+			copy(pageToken, result.NextPageToken)
+			// Make sure we don't include extraneous bytes if the new page token is
+			// shorter than the previous one (for whateve reason)
+			pageToken = pageToken[:nextPageTokenLen]
+		} else {
+			// No further results
+			moreResults = false
+		}
+
+		for _, elem := range result.Elements {
+			// Error occurred retrieving block metadata, use default values
+			blockStart := time.Unix(0, elem.Start)
+			if elem.Err != nil {
```

Comment:
This error handling doesn't seem very good, but this is how the existing code handles it

Reply:
I guess it kind of works because the block is so far back in the past (since elem.Start is 0?) that it just ignores the this one entirely? Seems like we should emit a stat or something